### PR TITLE
Dynamic Call Graph: Variable Type Analysis (VTA)

### PR DESCRIPTION
**Pull request by DavidBakerEffendi**
- Based on the pointer assignment graph (PAG) pass, will refine possible call targets using pointer information
- Limited to intra-procedural PAG information as the PAG is intra-procedural (for now)
- Does not consider `Operators.cast`
- If no pointer information is available it will default to CHA

### CODE REVIEW

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala [hunk 1]
```diff
@@ -134,6 +139,47 @@ class DynamicCallLinker(cpg: Cpg) extends SimpleCpgPass(cpg) {
     }
   }
 
+  /** Attempts to refine possible call targets using points-to information.
+    */
+  private def filterWithVariableTypeInformation(
+    call: Call,
+    tgts: mutable.LinkedHashSet[String]
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala:146]: I may be missing something, but it looks like `tgts` doesn't need to be mutable?

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala [hunk 2]
```diff
@@ -134,6 +139,47 @@ class DynamicCallLinker(cpg: Cpg) extends SimpleCpgPass(cpg) {
...
+      val thisType = receivers.collect { case x: Identifier => Option(x.typeFullName) }.flatten.toSet
+      filterTargets(tgts, thisType)
+    } else {
+      // Receiver is an object that may have points-to information, attempt to use it
+      val allocatedSuperTypes =
+        receivers
+          .flatMap(_.pointsToOut)
+          .collect {
+            case x: Call if x.methodFullName == Operators.alloc            => x.typeFullName
+            case x: Call if x.methodFullName == Operators.arrayInitializer => x.typeFullName
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala:165]:
```suggestion
            case x: Call if x.methodFullName == Operators.alloc || x.methodFullName == Operators.arrayInitializer => 
              x.typeFullName
```

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala [hunk 3]
```diff
@@ -134,6 +139,47 @@ class DynamicCallLinker(cpg: Cpg) extends SimpleCpgPass(cpg) {
...
 
+  /** Attempts to refine possible call targets using points-to information.
+    */
+  private def filterWithVariableTypeInformation(
+    call: Call,
+    tgts: mutable.LinkedHashSet[String]
+  ): mutable.LinkedHashSet[String] = {
+    // We can refine possible calls using the potential allocation site
+    val receivers      = call.receiver.l
+    val isThisReceiver = receivers.flatMap(x => Option(x.code)).exists(x => x.contains("this"))
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala:150]: I didn't ask the compiler, but shouldn't this be equivalent (and nicer)?
```
val isThisReceiver = receivers.code(".*this.*").nonEmpty
```

**Review by mpollmeier — COMMENTED** [review]
there's a few debug `println` statements in there, please either remove or replace with a logger. 
Apart from that looks fine from my side, however I'd rather also have @bbrehm 's eyes on this before merging

**Comment by DavidBakerEffendi** [comment]
> there's a few debug `println` statements in there, please either remove or replace with a logger.
> Apart from that looks fine from my side, however I'd rather also have @bbrehm 's eyes on this before merging

Yeah, sorry, as I have recently found a bug in the PAG pass that this merges into I have found a bug I need to fix there then I will merge that up in this branch and remove the `WIP` prefix

**Comment by DavidBakerEffendi** [comment]
A main question I have is currently I use `typeFullName` for VTA but should I perhaps look at `EVAL_TYPE` edges to be more flexible to dynamic languages?

**Comment by mpollmeier** [comment]
@bbrehm @ml86 can you answer that?

**Comment by DavidBakerEffendi** [comment]
> @bbrehm @ml86 can you answer that?

Is a fairly easy change, I've updated it to use `evalType` as the tests continue to pass and it will be able to leverage dynamic language type analysis passes if any.

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala [hunk 4]
```diff
@@ -134,6 +139,42 @@ class DynamicCallLinker(cpg: Cpg) extends SimpleCpgPass(cpg) {
     }
   }
 
+  /** Attempts to refine possible call targets using points-to information. Treats properties as nullable to avoid NPE.
+    */
+  private def filterWithVariableTypeInformation(call: Call, tgts: Set[String]): Set[String] = {
+    // We can refine possible calls using the potential allocation site
+    val receivers      = call.receiver.l
+    val isThisReceiver = receivers.filter(_.code != null).code(".*this.*").nonEmpty
```
[the remark this answers is not in the document]
Comment (DavidBakerEffendi) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala:147]: Something like the following seems to do the trick
```scala
val isThisReceiver = Traversal(receivers).isIdentifier.refsTo.collect {
      case p: MethodParameterIn if p.index == 0 => p
    }.nonEmpty
```

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala [hunk 5]
```diff
@@ -134,6 +139,42 @@ class DynamicCallLinker(cpg: Cpg) extends SimpleCpgPass(cpg) {
...
+      filterTargets(tgts, allocatedSuperTypes)
+    }
+  }
+
+  @inline
+  private def filterTargets(tgts: Set[String], allowedSet: Set[String]): Set[String] = {
+    if (allowedSet.isEmpty)
+      tgts // if allowed set is empty then we likely failed to receive points-to information
+    else
+      tgts.filter { t => allowedSet.contains(t.substring(0, t.lastIndexOf("."))) }
```
[the remark this answers is not in the document]
Comment (DavidBakerEffendi) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala:175]: `call.name` may be duplicated across types inheriting methods from parent types so `tgts` are `methodFullNames` which I have reflected in the following less-hacky update:
```scala
 tgts.filter { methodFullName =>
        methodMap(methodFullName).definingTypeDecl.fullName.headOption match {
          case Some(tgtTypeDecl) => allowedSet.contains(tgtTypeDecl)
          case None => false
        }
      }
```

[the remark this answers is not in the document]
Comment (ml86) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala]: Works.

**Comment by DavidBakerEffendi** [comment]
Closing for the same reason mentioned in #1249

**Pull request closed without merging**